daemon/daemon/docker_watcher.go

```diff
@@ -200,136 +201,92 @@ func (d *Daemon) updateProbeLabels(dockerID string) (bool, *types.Container, err
-		}
+	changed := false
 
-		if isProbe {
-			probeLabelsSHA256, err := ciliumContainer.OpLabels.ProbeLabels.SHA256Sum()
-			if err != nil {
-				log.Errorf("Error calculating SHA256Sum of labels %+v: %s", ciliumContainer.OpLabels.ProbeLabels, err)
-			}
-			if probeLabelsSHA256 != newLabelsSHA256 {
-				isNewContainer = true
-				epLabelsSHA256, err = ciliumContainer.OpLabels.EndpointLabels.SHA256Sum()
-				if err != nil {
-					log.Errorf("Error calculating SHA256Sum of labels %+v: %s", ciliumContainer.OpLabels.EndpointLabels, err)
-				}
-				// probe labels have changed
-				// we need to find out which labels were deleted and added
-				deletedLabels := ciliumContainer.OpLabels.ProbeLabels.DeepCopy()
-				for k, v := range newLabels {
-					if ciliumContainer.OpLabels.ProbeLabels[k] == nil {
-						tmpLbl1 := *v
-						tmpLbl2 := *v
-						ciliumContainer.OpLabels.AllLabels[k] = &tmpLbl1
-						ciliumContainer.OpLabels.EndpointLabels[k] = &tmpLbl2
-					} else {
-						delete(deletedLabels, k)
-					}
-				}
-
-				for k := range deletedLabels {
-					delete(ciliumContainer.OpLabels.AllLabels, k)
-					delete(ciliumContainer.OpLabels.EndpointLabels, k)
-				}
-			}
+	c.OpLabels.Orchestration.MarkAllForDeletion()
+	c.OpLabels.Disabled.MarkAllForDeletion()
+
+	for k, v := range newLabels {
+		if c.OpLabels.Disabled[k] != nil {
+			c.OpLabels.Disabled[k].DeletionMark = false
 		} else {
-			// If it is not probe then all newLabels will be applied
-			epLabelsSHA256, err = ciliumContainer.OpLabels.EndpointLabels.SHA256Sum()
-			if err != nil {
-				log.Errorf("Error calculating SHA256Sum of labels %+v: %s", ciliumContainer.OpLabels.EndpointLabels, err)
-			}
-			if epLabelsSHA256 != newLabelsSHA256 {
-				isNewContainer = true
-				ciliumContainer.OpLabels.EndpointLabels = newLabels
+			if c.OpLabels.Orchestration[k] != nil {
+				c.OpLabels.Orchestration[k].DeletionMark = false
+			} else {
+				tmp := *v
+				c.OpLabels.Orchestration[k] = &tmp
+				changed = true
 			}
 		}
-		cont = types.Container{
-			ContainerJSON: dockerCont,
-			OpLabels:      ciliumContainer.OpLabels,
-			NRetries:      ciliumContainer.NRetries,
-		}
 	}
 
-	if isNewContainer {
-		if err := d.DeleteLabelsBySHA256(epLabelsSHA256, dockerID); err != nil {
-			log.Errorf("Error while deleting old labels (%+v) of container %s: %s", epLabelsSHA256, dockerID, err)
-		}
+	if c.OpLabels.Orchestration.DeleteMarked() || c.OpLabels.Disabled.DeleteMarked() {
+		changed = true
 	}
 
-	d.containers[dockerID] = &cont
-	contCpy := cont
-
-	return isNewContainer, &contCpy, nil
+	return changed, c, nil
 }
 
-func (d *Daemon) updateContainer(container *types.Container, isNewContainer bool) error {
+func (d *Daemon) updateContainerLabels(container *types.Container, isNewContainer bool) error {
 	if container == nil {
 		return nil
 	}
 
 	dockerID := container.ID
 
-	secCtxlabels, isNewLabel, err := d.PutLabels(container.OpLabels.EndpointLabels, dockerID)
+	// Remove the old labels entry
```

Comment:
The logic to check for label changes is in `func (h *putEndpointIDLabels) Handle`